File: 46. Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<int> path;
        vector<vector<int>> res;
        DFS(res, path, nums);
        return res;
    }
    
    void DFS(vector<vector<int>>& res, vector<int>& path, vector<int>& nums)
    {
        if (path.size() == nums.size())
        {
            res.push_back(path);
            return;
        }
        for (int i = 0; i < nums.size(); i++)
        {
            if (find(path.begin(), path.end(), nums[i]) == path.end())
            {
                path.push_back(nums[i]);
                DFS(res, path, nums);
                path.pop_back();
            }
        }
    }
};

File: 46. Permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        self.res=[]
        self.DFS([], nums)
        return self.res
    
    def DFS(self, path, nums):
        if len(path)==len(nums):
            self.res+=[path[:]]
            return
        for i in xrange(len(nums)):
            if nums[i] in path:
                continue
            path+=[nums[i]]
            self.DFS(path, nums)
            path.pop()
